### PR TITLE
all: Add some generics and incorporate them into the codebase.

### PR DESCRIPTION
**Pull request by Shaptic**
<details>
  <summary>PR Checklist</summary>
  
### PR Structure

* [x] This PR has reasonably narrow scope (if not, break it down into smaller PRs).
* [x] This PR avoids mixing refactoring changes with feature changes (split into two PRs
  otherwise).
* [x] This PR's title starts with name of package that is most changed in the PR, ex.
  `services/friendbot`, or `all` or `doc` if the changes are broad or impact many
  packages.

### Thoroughness

* [x] This PR adds tests for the most critical parts of the new functionality or fixes.
* [x] I've updated any docs ([developer docs](https://developers.stellar.org/api/), `.md`
  files, etc... affected by this change). Take a look in the `docs` folder for a given service,
  like [this one](https://github.com/stellar/go/tree/master/services/horizon/internal/docs).

### Release planning

* [x] I've updated the relevant CHANGELOG ([here](services/horizon/CHANGELOG.md) for Horizon) if
  needed with deprecations, added features, breaking changes, and DB schema changes.
* [x] I've decided if this PR requires a new major/minor version according to
  [semver](https://semver.org/), or if it's mainly a patch change. The PR is targeted at the next
  release branch if it's not a patch change.
</details>

### What
Adds and incorporates a generic `Min`/`Max` rather than customs ones per-type and per-package.

### Why
blessed generics

### Known limitations
There might be other places we're doing this inline that are much harder to find, e.g.

```go
if a < b {
    a = b
}
```

one-offs that didn't warrant a dedicated `min` for that package.

### CODE REVIEW

File: support/generics/math.go [hunk 1]
```diff
@@ -0,0 +1,21 @@
+package generics
```
Comment (leighmcculloch) [inline support/generics/math.go:1]: `generics` is a reasonable generic name. Given that this package contains functions that work with ordered types, how about we named it `ordered`?

Comment (Shaptic) [reply support/generics/math.go:1]: The idea was to expand it over time to include other helpers (as you can tell from my ambitious branch naming). Min/Max are the most obvious offenders, but now that I've added and incorporated a `Set[T any]` container, would you reconsider the rename? :wink:

Comment (Shaptic) [reply support/generics/math.go:1]: I also plan to add both

```golang
func mapKeys[T any, U any](m map[T]U) []T      // and
func mapValues[T any, U any](m map[T]U) []U
```

once this gets merged into the `lighthorizon` branch.

Comment (leighmcculloch) [reply support/generics/math.go:1]: I'd put the map functions into a `collections/map` package, and the `Set` type into a `collections/set` package. i.e.
```
ordered
collections/set
collections/map
```

Comment (leighmcculloch) [reply support/generics/math.go:1]: This is a small detail and I defer to you. It's great we can do these things now 🎉.

**Review by leighmcculloch — APPROVED** [review]
👏🏻

File: support/generics/math.go (hunk 1 above)
Comment (Shaptic) [reply support/generics/math.go]: I guess my goal was to keep everything in its own subfolder (with no _explicit_ reason besides it being a 1.18+ feature and the general satisfaction of keeping a PR in a single folder), but I do like these names! Done in 4d58a361.

**Pull request merged**